presto-main/src/main/java/com/facebook/presto/operator/SimpleArrayAllocator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.ArrayAllocator;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.IdentityHashMap;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.util.Collections.newSetFromMap;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A simple {@link ArrayAllocator} that caches returned arrays.  When retrieving an array which is
+ * cached, this implementation will first discard all smaller arrays encountered to ensure amortized
+ * constant time access to the appropriately sized array.
+ */
+@NotThreadSafe
+class SimpleArrayAllocator
+        implements ArrayAllocator
+{
+    private static final int DEFAULT_MAX_OUTSTANDING = 1000;
+    private final int maxOutstandingArrays;
+
+    private final Deque<int[]> intArrays = new ArrayDeque<>();
+    private final Set<int[]> borrowedIntArrays = newSetFromMap(new IdentityHashMap<>());
+    private long estimatedSizeInBytes;
+
+    public SimpleArrayAllocator()
+    {
+        this(DEFAULT_MAX_OUTSTANDING);
+    }
+
+    public SimpleArrayAllocator(int maxOutstandingArrays)
+    {
+        if (maxOutstandingArrays <= 0) {
+            throw new IllegalArgumentException("maxOutstandingArrays must be positive");
+        }
+        this.maxOutstandingArrays = maxOutstandingArrays;
+    }
+
+    @Override
+    public int[] borrowIntArray(int positionCount)
+    {
+        checkState(borrowedIntArrays.size() < maxOutstandingArrays, "Requested too many arrays: %s", borrowedIntArrays.size());
+        while (!intArrays.isEmpty() && intArrays.peek().length < positionCount) {
+            int[] popped = intArrays.pop();
+            estimatedSizeInBytes -= sizeOf(popped);
+        }
+        int[] array;
+        if (intArrays.isEmpty()) {
+            array = new int[positionCount];
+            estimatedSizeInBytes += sizeOf(array);
+        }
+        else {
+            array = intArrays.pop();
+        }
+        checkState(borrowedIntArrays.add(array), "Attempted to borrow array which was already borrowed");
```

Comment:
This is not a caller's problem, but rather a bug in the implementation of this class. Hence, `verify` would be more appropriate.